### PR TITLE
Bullet & ODE heightmaps

### PR DESCRIPTION
**Pull request by JenniferBuehler**
This adds support for heightmaps with Bullet or ODE collision detectors.
I'm also planning to add FCL support by using the Octomap, but I think this should be a separate PR.

There are a couple of things yet to discuss and agree on before this could be merged.

**1. Flipping Y values for bullet**

I  still don't find the solution for flipping the Y values proposed here very pretty (flipping y is required for bullet so that it has the same pose as the heightmap in ODE). I've thought about a couple of different ways how to to do this, but all of them have something undesirable about it. The main issue is that ``dynamics::HeightmapShape`` is built (possibly externally) before the bullet collision object (and with it the bullet ``btHeightfieldTerrainShape``) is created from it. If we create and fill ``dynamics::HeightmapShape`` externally (e.g. from within the Gazebo [DARTHeightmapShape](https://bitbucket.org/JenniferBuehler/gazebo/src/1e8ad5554a8a82f658462574add4908f2cb3b8ea/gazebo/physics/dart/DARTHeightmapShape.cc#lines-65)), we would need to know at this point whether the Y values need to be flipped or not, so we need to know which collision detector is used. However ``dynamics::HeightmapShape`` cannot provide this information; we could instead query ``World`` for the used ``CollisionDetector``, and if it's bullet, then flip the Y values. But  this imposes too much required knowledge about the internals for the user, and it also requires a pointer to the dart ``World``, so I don't consider this a good idea. It's best if the flipping of Y values happens automatically "behind the scenes". Which brings up the next issue: ideally we'd like to have the option to use an existing height value buffer and pass it into [dynamics::HeightmapShape::setHeightField()](https://github.com/JenniferBuehler/dart/blob/659564d18f3de47bc70fc3de8dee7b5d626b71e4/dart/dynamics/HeightmapShape.hpp#L96) (at the moment, a local copy of the vector is kept, which can be a waste of memory if the caller aims to keep their height field vector as well -that's the case for gazebo's [HeightmapShape](https://bitbucket.org/JenniferBuehler/gazebo/src/dart_heightmap_with_bullet/gazebo/physics/HeightmapShape.hh)). But modifying an external caller's buffer data is not such a good idea. I'm not quite sure which solution would be more desirable, they are both not so ideal. So for now, I am just copying the height values vector, and flipping the y values in [BulletCollisionDetector](https://github.com/JenniferBuehler/dart/blob/659564d18f3de47bc70fc3de8dee7b5d626b71e4/dart/collision/bullet/BulletCollisionDetector.cpp#L933) (see also comment in the code linked).

Any thoughts on this? Which solution would you consider more suitable, or you have other ideas?

**2. Relative transforms**

Bullet moves the origin of the heightmap to the center of its AABB (see also [comment in btHeightfieldTerrainShape](https://github.com/bulletphysics/bullet3/blob/ab4b6638002c86d7d0c5d4a1b8e5728a227a92e3/src/BulletCollision/CollisionShapes/btHeightfieldTerrainShape.h#L44)). This is different to ODE, so again we need to make the behavior uniform. What's not so lovely is that [BulletCollisionDetector::createBulletCollisionShape()](https://github.com/JenniferBuehler/dart/blob/659564d18f3de47bc70fc3de8dee7b5d626b71e4/dart/collision/bullet/BulletCollisionDetector.cpp#L406) only returns a ``btCollisionShape``, which has no information about a geometry's permanent relative transform to its parent object. So this relative transform has to be returned additionally from [where it is created](https://github.com/JenniferBuehler/dart/blob/659564d18f3de47bc70fc3de8dee7b5d626b71e4/dart/collision/bullet/BulletCollisionDetector.cpp#L955).  For now I've solved this extra return value via an output parameter in the affected functions. This relative transform is then "dragged around" as parameter until it is remembered in [BulletCollisionObject](https://github.com/JenniferBuehler/dart/blob/659564d18f3de47bc70fc3de8dee7b5d626b71e4/dart/collision/bullet/BulletCollisionObject.cpp#L68), where the relative transform can then always be applied when the shape is re-positioned.    

What's not so nice about this: We only need this extra relative transform for the heightmap shape, all other shapes don't use it. Although maybe in future there will be other shapes for which relative transforms have to be applied to the geometry in order to make it uniform with how the same shape is handled in other collision detectors (?).

Similar issue for ODE, which uses Y as up axis instead of Z and therefore requires a rotation, but for ODE I found a solution which is less invasive (temporarily remembering the relative transform in the geometry itself, until a permanent geometry offset can be set after it is assigned to the body, which overwrites the geometry transform).

Is the proposed solution OK, or is it totally not desirable to keep an extra ``btTransform`` in ``BulletCollisionObject``?

**3. Target branch**

I'm not sure which branch I should submit this PR to, I've just used ``release-6.5`` for now. It would be good to have this available for Gazebo ASAP, see also implementation in the [dart_heightmap_with_bullet branch](https://bitbucket.org/JenniferBuehler/gazebo/src/dart_heightmap_with_bullet/) for Gazebo.

### CODE REVIEW

**Comment by JenniferBuehler** [comment]
I'll have to check about the test failure, not sure what it could be, it's passing on my system.

**Comment by JenniferBuehler** [comment]
Never mind about my last comment, I only was testing my added test which passes, but failures are due to issue #717 ;)

**Comment by JenniferBuehler** [comment]
See also [gazebo pull request 2956](https://bitbucket.org/osrf/gazebo/pull-requests/2956/).

I have used gazebo to visually check the results, it all looks fine both with ODE and bullet. 

It would be nice to have a visualization of the Heightmap shapes in DART as well, however we'd have to decide how to do this, e.g. reading from image file and which libraries to use etc. I'm happy to add this once I know your opinion on what would be the most desirable way to go about this.

**Comment by jslee02** [comment]
Thank you for the PR! I'll take a look at this PR this weekend.

Regarding target branch, I switched it to `release-6.6` because I'm planning to release 6.5.0 today. Once this PR is merged, I'll release 6.6 with this change so that this can be used in Gazebo ASAP.

**Comment by JenniferBuehler** [comment]
Hi JS, thanks! Have a lovely weekend :)

**Comment by JenniferBuehler** [comment]
Hi JS,
thanks for looking at this! Now it's me who doesn't get around to look at this, I'm currently at a fair and will then be on holiday, so it will take a couple of weeks until I get around to it.
Cheers!

File: dart/collision/ode/detail/OdeHeightmap.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  assert(width >= 2);
+  assert(height >= 2);
+  if ((width < 2) || (height < 2))
+  {
+    dtwarn << "Cannot create height field of dimensions " << width
+           << "x" << height << ", needs to be at least 2" << std::endl;
+    return;
+  }
+
+  dGeomHeightfieldDataBuildSingle(
```
[the remark this answers is not in the document]
Comment (JenniferBuehler) [reply dart/collision/ode/detail/OdeHeightmap.cpp]: Yes, indeed. I must have updated this function with copy and paste from the single precision version... oupsies! Thanks for finding this.

File: unittests/comprehensive/test_Collision.cpp [hunk 2]
```diff
@@ -1112,6 +1112,291 @@ TEST_F(COLLISION, testPlane)
...
+  EXPECT_EQ(shape->getHeightField()[2], heights1[0]);
+  EXPECT_EQ(shape->getHeightField()[3], heights1[1]);
+
+  // test with odd number of rows
+  std::vector<HeightmapShape::HeightType> heights2 = {-1, -2, 3, 3, 2, 1};
+  shape->setHeightField(2, 3, heights2);
+  shape->flipY();
+  EXPECT_EQ(shape->getHeightField()[0], heights2[4]);
+  EXPECT_EQ(shape->getHeightField()[1], heights2[5]);
+  EXPECT_EQ(shape->getHeightField()[2], heights2[3]);
```
[the remark this answers is not in the document]
Comment (JenniferBuehler) [reply unittests/comprehensive/test_Collision.cpp]: Yes :o

File: dart/dynamics/HeightmapShape.hpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+private:
+  /// \brief scale of the heightmap
+  Eigen::Vector3d mScale;
+
+  /// \brief width (x) and depth (y) of height field
+  size_t mWidth, mDepth;
+
+  /// \brief height field. Must be of size mWidth*mDepth.
+  mutable std::vector<HeightType> mHeights;
```
[the remark this answers is not in the document]
Comment (JenniferBuehler) [reply dart/dynamics/HeightmapShape.hpp]: Yes that makes sense, I haven't thought about using Eigen types actually. Will add this.

**Comment by JenniferBuehler** [comment]
Thanks for looking at this, and sorry for the late response.

I've now changed it to Eigen::Matrix in the last commit.

Regarding your comments:

1) **Flipping Y values**: I'm not sure which convention is more widely used either. In Gazebo, the y values are also flipped for bullet only, that's why I also did it here for bullet instead of for ODE. Not sure if there is a convention at all... also whether z or y is used as up axis varies. +z axis up is the most widely used standard, and so one could also argue that the values should be increasing row-wise (rows along +x) in +y direction, so everything goes towards +. However one could also argue that it should be -y instead (like it is interpreted now), because in image processing the rows are read top-down. This is how ODE understands it. Makes sense too, because the height fields are read from images to begin with. I don't mind either way, which do you prefer? I can change it.    
As for your idea with the ``GetYFlippedHeightField()`` function: also, good point. Side effect is that if we add a ``GetYFlippedHeightField()`` (and hence never modify the buffer in ``dart::HeightmapShape``), then we also have to store yet another copy of the height field with the flipped values for bullet, because ``btHeightfieldTerrainShape`` takes a pointer to the data buffer (I'm pretty sure it doesn't copy it again but admittedly haven't checked the source, the documentation doesn't mention it). So in the example of using this in Gazebo, there will then be 1) the height field data in ``physics::HeightmapShape``, and 2) there is the copied data in ``dart::HeightmapShape``, and then we would 3) additionally need the flipped Y values for bullet stored locally somewhere. That's not exactly awesome ;) I suggest instead to add a flag to ``dart::HeightMapshape`` indicating whether the y values are flipped or not, and then the user just has to un-flip the values if they want the original data. In ``BulletCollisionDetector`` where ``flipY()`` is  called, we'd then only flip if the data is not already flipped. What do you think about this solution?

2) **Relative transforms**: Yes that is possible. I've made the initial implementation as removed from modified buffers as possible, but now that we pretty much decided to always keep that copy in ``dart::HeightmapShape`` anyway, we can do that.
However if for any reason some other shapes (e.g. other collision engines or shapes added in future) have relative transforms, we have to do the same thing again... plus, what's more annoying, similar as in point (1) we will have to keep track of locally applied transforms to the data as well. I personally prefer the option in which local transforms are possible for all shapes in general (even if so far only used for the height map), local transforms may be necessary e.g for cylinders which sometimes have the origin at the base, sometimes in the middle, and sometimes the up-axis varies too. But I understand if you'd like to keep this PR as light as possible. So we can either keep it as-is, including transforms, or I can do the transform of the modified heightmap for bullet only. Let me know what you prefer.

Thanks again for looking into this!

**Comment by JenniferBuehler** [comment]
Just checked the CI and saw that there is a new failure with the changes for the y-values flipping, hang in there while I fix this.

**Comment by JenniferBuehler** [comment]
Found the issue, actually it's a bit annoying. Eigen doesn't store the data the same way as std::vector does, so in the end we can't do ``heights.data()`` to pass the data into ``btHeightfieldTerrainShape``, but have to transform the data first (see implementation in ``HeightmapShape::getHeightFieldAsVector()``). Are you sure you would much prefer the Eigen::Matrix version just for the simplified ``flipY()``?

**Comment by jslee02** [comment]
Thanks for the updates!

> Found the issue, actually it's a bit annoying. Eigen doesn't store the data in the same order as required for the height values buffer (i.e. returned from data() in std::vector).

Quick answer: Eigen supports two kinds of storage orders: [column major and row major](https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html). So `Eigen::Matrix<HeightType, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>` would be helpful. I'm reading the code why `getHeightFieldAsVector()` is necessary.

**Comment by JenniferBuehler** [comment]
Awesome, I didn't know about ``Eigen::RowMajor``, I've only used Eigen matrices for maths before... thanks! That fixed it.

Ok so if you want to use the height field data for anything else to perform some calculations on it, then it may be helpful to have it as Eigen::Matrix.

**Comment by JenniferBuehler** [comment]
Still getting the following failures:

```
[  FAILED  ] COLLISION.SimpleFrames
[  FAILED  ] COLLISION.BoxBox
[  FAILED  ] COLLISION.Filter
[  FAILED  ] COLLISION.CreateCollisionGroupFromVariousObject
```

I first thought this is this due to issue #717, but a diff to the ``glut_macos`` branch which builds in pull request #1082 shows no relevant differences in this regard.
Do you know why there is this difference in the CI build?

I'm also unusre why Codacity uses an old version of ``test_Collision.cpp`` (I removed the const-ness from the variable, which is a variable one can play around with to test. So the values shouldn't change during runtime, but Codacity doesn't like it being used in the conditional expression).

I haven't got much experience with the automated check tools, sorry if those are silly questions ;)

**Comment by jslee02** [comment]
Those failures are due to that `btTransform` isn't set by [calling this function](https://github.com/dartsim/dart/pull/1069/files#diff-a1d1fe81bb515a1eda7e4905f49f9615R337). This is the case that a `btCollisionShape` instance was created from the same DART `ShapeFrame`; the function just returns the same `btCollisionShape` without setting `relativeTransform` under the current implementation.

I fixed this in [this branch](https://github.com/dartsim/dart/tree/bullet_heightmap_js) (forking it out of your branch). Let me create a PR of it targeting to your branch by hopefully today.

**Comment by jslee02** [comment]
@JenniferBuehler I've create a PR (https://github.com/JenniferBuehler/dart/pull/1) that includes some fixes and updates targeting to be merged into this PR. Please take a look at it when you have a chance.

**Comment by JenniferBuehler** [comment]
Thanks for the PR.
Now all tests are passing, except the heightmapBox test is still failing for double, I'll fix this up later because I gotta run now.

**Comment by JenniferBuehler** [comment]
Ok, so double is actually not supported in bullet, which only supports float, unsigned char and float. This is not clearly explained in the documentation, but you can see it in the [bullet source here](https://github.com/bulletphysics/bullet3/blob/3d5c92149578fc5ff82327b8251cc3f0ee877fca/src/BulletCollision/CollisionShapes/btHeightfieldTerrainShape.cpp#L152).

I have removed the support of double height fields for bullet, and also got rid of the test for it. It would be nice to automatically convert a double field to float, but then again we have to maintain duplicate storage of the height field (the question would also be where to store this)... so I think altogether it's better to just not support this, or what do you think?

**Comment by jslee02** [comment]
> It would be nice to automatically convert a double field to float, but then again we have to maintain duplicate storage of the height field (the question would also be where to store this)... so I think altogether it's better to just not support this, or what do you think?

I'm happy with the current implement. The automatic conversion can be revisited later if it's necessary.

Also, the current implementation about relative transform looks good. My last concern is the flipping Y. The current implementation can be a problem when we use the shape height map shape in different collision detectors (e.g., ODE and Bullet) when the collision detector have different convention on how to store the height data.

Here is my proposal:

The idea is supporting the two convention explicitly by adding a few more public API to `HeightmapShape` such as `getHeightField1()`/`getHeightField2()` and `setHeightField1()`/`setHeightField2()` (names are TBD). This means basically the shape should maintain two data for the two conventions. The motivation of doing this is to get the right data for queries with different conventions. Of course this can be inefficient for example when the height data is set by `setHeightField1()` and the data is extracted by `getHeightField2()`, which requires data conversion from convention 1 to convention 2 internally. But I think it's the users responsibility to be consistent in setting and getting the height data.

We can additionally provide cleaning functions that deallocates the data of unused convention. If the user knows the data of a particular convention won't be used, then the data can be deleted by calling the function. Note that the cleaning functions shouldn't remove the data if the data is the only one (we need to hold at least one data of a convention so that it can be converted into other convention).

If this sounds good then I'll add this functionality. Let me know what you think.

Caveat: Please consider that I wrote this in motion so the explanation can be unclear and can contain many errors.

**Comment by JenniferBuehler** [comment]
Hi JS,

I see what you mean with the ``HeightmapShape`` flipped Y values. Yes if the data is used in different contexts or even different collision engines, then the buffer in ``HeightmapShape`` should not be manipluated. Good point!

I agree with your idea to maintain duplicate data only if two conventions are used at the same time. So if I understood this correctly, the flipped-y version of the height field will only be generated once ``getHeightfield2()`` has been called?

I'm not quite sure how you envision the automatic cleanup function, but it sounds good to have one :) So do you mean if the user subsequently to calling ``getHeightfield2()`` calls ``cleanUnsuedData()``, and nobody has called ``getHeightfield1()`` in the meantime either, then ``heightfield1`` will be deleted, instead of ``heightfield2``? Then we'd have to keep track of calls to the functions, not sure if that's very pretty...

In terms of names, I would maybe suggest to add a parameter ``conventionID`` or so, i.e. ``getHeightfield(uint8 conventionID)``... this could also be an enum (e.g. HEIGHTFIELD_REGULAR, HEIGHTFIELD_Y_FILPPED, etc). This means that if in future any other conventions are added, it could easily be integrated, instead of then having a ``getHeightfield3()`` or so. The height fields themselves can then be stored in an array indexed by the convention IDs. 
But that's only one idea, I'm happy with two names too.

If you're happy to add this to the PR, it would be awesome, then you can do the solution you would like to have right away :)

**Comment by jslee02** [comment]
> the flipped-y version of the height field will only be generated once getHeightfield2() has been called?

Yes.

> So do you mean if the user subsequently to calling getHeightfield2() calls cleanUnsuedData(), and nobody has called getHeightfield1() in the meantime either, then heightfield1 will be deleted, instead of heightfield2? Then we'd have to keep track of calls to the functions, not sure if that's very pretty...

I agree. That is one possible implementation but I'm trying to find a better way. My initial thought was that the cleanup functionality isn't automatic, but the cleanup function should be called by the caller. Then the automation is done by collision detector's claim/reclaim mechanism. However, this isn't also not that clean I think. Let me think about this further.

In the meantime, I think this PR is good to go. @mxgrey Please let me know if you'd like to review this PR further.

**Comment by JenniferBuehler** [comment]
Awesome, thanks for merging this :) I agree that the issue with the Y-values should be re-visited later, so that ``HeightmapShape`` can be re-used. For now I'm happy that the PR has been merged so that the Gazebo PR can move forward too. Thanks!

**Pull request merged**